src/main/java/com/google/sps/servlets/ManageListServlet.java

```diff
@@ -55,31 +59,68 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     String type = request.getParameter("type").toLowerCase();
     String action = request.getParameter("action").toLowerCase();
 
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
     // creates a key based on the passed in recipe ID
     Key key = KeyFactory.createKey("Recipe", recipeID);
     // creates a user based on the passed in user id token
     User user = new User(userIdToken);
 
-    // adds created key to appropriate list based on passed in list type
+    // adds or removes created key to/from appropriate list based on passed in 
+    // list type and passed in action
     if (type.equals("cookbook")) {
-      actionOnPlanner(user, key, action);
-    } else if (type.equals("planner")) {
       actionOnCookbook(user, key, action);
+    } else if (type.equals("planner")) {
+      actionOnPlanner(user, key, action);
     } else {
       System.out.println("AddToListServlet: invalid type " + type);
     }
   }
 
-  public void actionOnPlanner(User user, Key key, String action) {
+  public void actionOnCookbook(User user, Key key, String action) {
     if (action.equals("add")) {
       user.addCookbookKey(key);
+      increasePopularity(key);
     } else if (action.equals("remove")) {
       user.removeCookbookKey(key);
+      decreasePopularity(key);
     } else {
       System.out.println("ManageListServlet: invalid action " + action);
     }
   }
-  public void actionOnCookbook(User user, Key key, String action) {
+
+  public void increasePopularity(Key key) {
```

Comment:
My only  comment here is I would've used a general method like `updatePopularity` which could take a positive or negative int and would have a lot of the shared logic of the increase/decrease popularity functions, and then have your functions make call to that function. Just an as abstraction since the only apparent difference is the value change and the logic of `if (popularity < 0) { popularity = 0}`

Reply:
done